### ISSUE TITLE
feat: add step-count and loss history logging hooks for experiment tracking

### ISSUE THREAD

**Issue by kovacoj** [kovacoj1/optimizers#72]
## Problem

For a numerical mathematician running experiments, the most basic diagnostic is watching how the loss evolves over steps. Currently:

- Newton returns `None` — no way to track progress at all.
- LM returns `float`, EKF/KF return tensors, Annealing/Metropolis/Genetic return tensors — inconsistent and requires the user to manually maintain a loss history list.
- There's no way to log per-step diagnostics (step count, loss, key parameters like `mu` or `T`) without wrapping every `step()` call.

This is a significant usability gap for the target audience. A mathematician wants to:
1. Plot convergence curves.
2. Compare optimizers on the same problem.
3. Detect stagnation or divergence early.
4. Understand *why* an optimizer is or isn't converging (e.g., is `mu` exploding? is `T` decaying too fast?).

## Proposal

Add an optional callback or logging hook to the base optimizer interface. A minimal approach:

```python
optimizer = Newton([x], on_step=lambda info: print(info))
# info = {"step": 5, "loss": 0.0032, "mu": 1.5, ...}
```

Or a simpler approach: just make all `step()` methods return a consistent dict:

```python
return {"loss": loss, "step": self._step_count}
```

This is additive and non-breaking — existing code that ignores the return value is unaffected.

**Comment by kovacoj** [comment]
## Proposed fix — minimal, non-breaking approach

Standardize all `step()` return values to include a loss, and add an optional `_step_count` attribute for experiment tracking:

**1. Make Newton return the loss:**

```python
# Before
def step(self, closure):
    ...
    self.update_weights(direction)
    return

# After
def step(self, closure):
    ...
    self.update_weights(direction)
    with torch.no_grad():
        return closure().item()
```

This makes Newton consistent with the other optimizers — every `step()` now returns a scalar loss that can be used for tracking.

**2. Add `_step_count` to all optimizers:**

```python
class Newton(torch.optim.Optimizer):
    def __init__(self, ...):
        ...
        self._step_count = 0

    def step(self, closure):
        ...
        self._step_count += 1
        return loss
```

Users who want a loss history can simply:
```python
losses = []
for _ in range(100):
    losses.append(optimizer.step(closure))
```

**3. (Optional, separate issue) Add a callback hook:**

```python
class Newton(torch.optim.Optimizer):
    def __init__(self, ..., on_step=None):
        self._on_step = on_step

    def step(self, closure):
        ...
        if self._on_step:
            self._on_step({"step": self._step_count, "loss": loss})
        return loss
```

**Why not a more complex solution:** The target audience (numerical mathematicians) values simplicity and transparency. A callback hook is nice but adds complexity. The most impactful change is making Newton return the loss (currently `None`) so every optimizer has a usable return value. The `_step_count` is a close second. Both are additive, non-breaking changes.

**Breaking change note:** Newton currently returns `None`. Changing it to return `float` is technically a type change, but since `None` is unusable as a loss value, no reasonable code depends on it. Still worth noting in release notes.

**Comment by kovacoj** [comment]
Closing this as not planned for now.

After tightening the correctness/test/refactor issues, this looks more like a speculative API expansion than a necessary backend fix. Adding callbacks, internal histories, or dict return values would touch every optimizer and add cross-cutting surface area that makes the package less simple for numerical users.

The cleaner pattern for this project is to keep experiment tracking external:

```python
losses = []
for _ in range(100):
    loss = optimizer.step(closure)
    losses.append(loss)
```

For optimizers whose return type is not the desired plotting scalar, users can evaluate the closure immediately after `step()` and store the scalar they want. That keeps the optimizer code focused on numerical updates rather than logging policy.

If we later want a narrower change, I would file it separately as something concrete like "make Newton.step return final loss" rather than adding a global callback/history API.
